src/framework/emqttb_autorate.erl

```diff
@@ -16,30 +16,156 @@
+        { id        :: atom()
+        , parent    :: reference() | undefined
+        , current   :: float()
+        , conf_root :: atom()
+        , error     :: fun(() -> number())
+        , last_t    :: integer()
+        , last_err  :: number()
+        }).
+
+init(Config = #{id := Id, conf_root := ConfRoot, error := ErrF}) ->
+  MRef = case Config of
+           #{parent := Parent} -> monitor(process, Parent);
+           _                   -> undefined
+         end,
+  %% Create metrics:
+  emqttb_metrics:new_gauge(?AUTORATE_RATE(Id),
+                           [ {help, <<"Controlled value">>}
+                           , {labels, [id]}
+                           ]),
+  [emqttb_metrics:new_gauge(?AUTORATE_CONTROL(Id, I),
+                            [ {help, <<"Control output delta">>}
+                            , {labels, [id, term]}
+                            ]) || I <- [sum, p, i]],
+  %% Init PID:
+  Min = my_cfg(ConfRoot, [min]),
+  Current = maps:get(init_val, Config, Min),
+  Err = ErrF(),
+  set_timer(),
+  {ok, #s{ id        = Id
+         , parent    = MRef
+         , current   = Current
+         , conf_root = ConfRoot
+         , error     = ErrF
+         , last_err  = Err
+         , last_t    = os:system_time(millisecond)
+         }}.
+
+handle_call(get_counter, _From, S) ->
+  {reply, emqttb_metrics:gauge_ref(?AUTORATE_RATE(S#s.id)), S};
+handle_call(_, _From, S) ->
+  {reply, {error, unknown_call}, S}.
+
+handle_cast(_, S) ->
+  {noreply, S}.
+
+handle_info(tick, S) ->
+  set_timer(),
+  {noreply, update_rate(S)};
+handle_info({'DOWN', MRef, _, _, _}, S = #s{parent = MRef}) ->
+  {stop, normal, S};
+handle_info(_, S) ->
+  {noreply, S}.
+
 %%================================================================================
 %% Internal functions
 %%================================================================================
+
+
+update_rate(S = #s{ last_t    = LastT
+                  , error     = ErrF
+                  , current   = CO0
+                  , last_err  = LastErr
+                  , conf_root = ConfRoot
+                  , id        = Id
+                  }) ->
+  %% 0. Get current error
+  T = os:system_time(millisecond),
+  Dt = (T - LastT) / timer:seconds(1),
+  Err = ErrF(),
+  %% 1. Get config
+  Min = my_cfg(ConfRoot, [min]),
+  Max = my_cfg(ConfRoot, [max]),
+  MaxSpeed = my_cfg(ConfRoot, [speed]),
+  Kp = my_cfg(ConfRoot, [k_p]),
+  Ti = my_cfg(ConfRoot, [t_i]),
+  %% 2. Calculate the control output
+  De = LastErr - Err,
+  CP = Kp * De, %% Note: delta t is reduced here
+  CI = Kp * Err / Ti * Dt,
+  DeltaC = clamp(CP + CI, MaxSpeed * Dt),
+  CO = clamp(CO0 + DeltaC, Min, Max),
+  %% 3. Update metrics
+  emqttb_metrics:gauge_set(?AUTORATE_RATE(Id), round(CO)),
+  emqttb_metrics:gauge_set(?AUTORATE_CONTROL(Id, sum), round(DeltaC / Kp)),
+  emqttb_metrics:gauge_set(?AUTORATE_CONTROL(Id, p), round(CP / Kp)),
+  emqttb_metrics:gauge_set(?AUTORATE_CONTROL(Id, i), round(CI / Kp)),
```

Comment:
Q: since only `?AUTORATE_RATE(Id)`'s value ends up multiplied by `Kp`, is there a reason to multiply the other values by it before cancelling it by division later?  Is it to avoid some float arithmetic error?

Reply:
These metrics are reported just for grafana. Kp tends to be tiny, so if we report them as is, the graphs won't be very informative.